The fishers_validate_positions.py script provided here carries out the Fishers exact test on select
SNP positions (provided in the script).

The input required for this script is:

1. A filtered fasta file containing the alignments of all the isolates in the validation set.
2. A file containing the associated SNP positions corresponding to the file above.
3. A file containing the isolate names and phenotype severated by a tab. A label at the top of the
   file is also required.

The base forms of these files are generated by the SNP calling pipeline and then filtered by the script
described in the preparation directory (one level up from the current directory).

A file name is required to write out a biallelic position/SNP file.

Usage: python fishers_validate_positions.py fasta_alignments positions_file phenotype_file output_datafile
